src/Plugin/RulesAction/DataConvert.php

```diff
@@ -104,4 +104,13 @@ public function execute() {
     $this->setProvidedValue('conversion_result', $result);
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  public function refineContextDefinitions() {
+    if ($type = $this->getContextValue('target_type')) {
+      $this->pluginDefinition['provides']['conversion_result']->setDataType($type);
+    }
```

Comment:
A before/after test should be added to show that refining the type actually creates a change in type.